### ISSUE TITLE
launcher: systematic start error + 1s restart causes high CPU usage

### ISSUE THREAD

**Issue by mpl** [perkeep/perkeep#1200]
For an instance deployed on GCE with our launcher, the systemd service automatically restarts the service after 1s, in case of a failure to start.
It means that if Perkeep systematically fails to start, it will uselessly retry to start every second, which leads to high CPU usage for nothing.

**Comment by mpl** [comment]
As https://github.com/systemd/systemd/issues/6129 seems to suggest, I don't think there's any way to do that for now.

So maybe for now increase the restart to a more conservative value, like 10s, to alleviate the problem?

**Comment by euank** [comment]
I took a glance at the `pkg/deploy/gce/deploy.go` code and saw quite a few things I'd want to fix there.

One of them is moving from `cloud-config` to Ignition since, as [documented](https://coreos.com/os/docs/latest/cloud-config.html), cloud-config is both deprecated and broken.

I see it's also using `systemd-docker` to wrap `docker run`, which is also an unmaintained project by now.

In any case, I'd be happy to go through and clean that up, and get this issue along with it.

**Comment by euank** [comment]
After looking a little more, I strongly suspect the network + cpu usage is due to the `ExecStartPre` of `curl $camlistored | gunzip`...

That only really needs to be done once, but how the systemd unit is written, it's happening every time the unit restarts.

I don't think the right fix is changing the restart value or how it backs off, but to make it not do a `curl | gunzip` unless it needs to.

**Comment by mpl** [comment]
the systematic curl + gunzip is the easiest way to ensure automatic updates.
If you want to do without it, you still have to make sure in some other way that if we've uploaded a new image, they automatically use it.

Also, I wouldn't worry too much about fixing old and deprecated stuff, as we might move to a different environment sooner or later.

**Comment by euank** [comment]
> the systematic curl + gunzip is the easiest way to ensure automatic updates.

It's easy enough to `curl -z`, or save the gcs etag, or something of the like... and it's even easier to not gunzip if the file hasn't changed and it's already loaded into docker.

Unless we're sure that stuff is vanishing *very* soon, it still seems worth fixing (after all, this issue was filed based on a user report).

**Comment by bradfitz** [comment]
@euank, yeah, ideally we'd be much smarter about HTTP caching & only nuke the output directory & untar things if the HTTP origin had changed.

Somebody could steal some code from my https://github.com/golang/build/blob/master/internal/httpdl/httpdl.go

**Comment by euank** [comment]
Thinking about it a little more, for the download part, `docker pull` from a registry  (e.g.  gcr.io) already deals with only downloading changes, making sure the image in docker is up to date,  etc etc.

Is there any reason not to switch to that?

**Comment by bradfitz** [comment]
Yeah, totally. Doing a pull from gcr.io would be best. I think it was kinda new (if it even existed?) when we started this.

**Comment by euank** [comment]
I looked into this a little more and posted a WIP CL here: https://perkeep-review.googlesource.com/c/perkeep/+/18046

One of the remaining changes is to update any release documentation or automation for updating that gcr image.

I'd appreciate any pointer to existing documentation or automation around that.

**Comment by mpl** [comment]
I've started moving the launcher to Container-Optimized OS (which was the other environment I mentioned above) yesterday, which as far as I know uses cloud-config, and not ignition. So I don't think it's very useful to switch to ignition now since we're going to go back to cloud-config right after.

That said it seems there's other useful things in your CL, I'm going to have a look.

**Comment by euank** [comment]
@mpl I don't see any issue or other discussion about such a transition. I'm curious what the motivation is. Assuming perkeep's launcher code may intend to eventually support other clouds, like AWS, it seems odd to pick an OS which is not available anywhere else.

Am I missing discussion somewhere or other broader context?

**Comment by mpl** [comment]
@euank no, you're not missing any previous discussion about it afaik.

I'll let @bradfitz answer that one.

**Comment by bradfitz** [comment]
Where the world is heading and where we'd like to be ourselves is only deploying a container, not a VM.

All the cloud providers are working on container-only offerings where you don't need to think about the underlying VMs. They're not entirely there yet, but that's the direction.

Using Container-Optimized OS is a stopgap on GCE. It lets you only think about the container(s) as the unit of deployment, but GCE implements it (for now) with a VM dedicated to running your container(s).

Later that'll go away.

Likewise, on AWS and Azure and other places we'll use the exact same Docker/OCI/etc container image and run it on all clouds.

**Comment by euank** [comment]
@bradfitz I'm sold on the container as a unit of deployment in the future for this and that dealing with VMs at all is a stop-gap.

I don't see how any of that makes COS a better stop-gap than the existing choice of CoreOS though.

> Using Container-Optimized OS is a stopgap on GCE. It lets you only think about the container(s) as the unit of deployment

It does so just as well as CoreOS does. Fundamentally, COS and CoreOS are basically systemd + dockerd.. The only major differences are CoreOS using Ignition (woo) and more memory due to its current docker configuration.
For the single node perkeep's deploying, I think their update model and management is effectively identical... It just happens CoreOS is what's used already and is slightly easier to deploy to due to Ignition, so to repeat, I don't see a reason swapping in COS helps.

Once we're free of the VM (be it by using GKE, or a fargate like service gce releases, or whatever), I don't think the transition will be in any way different from COS or CoreOS to said new thing.

So again, I'd like to ask, what's the point in switching to COS? I get why neither COS nor CoreOS is the future, I don't get why there's any benefit shifting between them in the present.

**Comment by bradfitz** [comment]
The Go build uses COS so I'm more familiar with it. I lost my familiarity with CoreOS a few years ago and even then it wasn't very container-friendly. 

I just looked at your CL and I see a pile of shell-in-yaml goo at https://perkeep-review.googlesource.com/c/perkeep/+/18046/1/pkg/deploy/gce/deploy.go#889 that seems like something I'd never want to look at or maintain.

COS manages containers. It seems that CoreOS still manages systemd units and it's our job to do all the Docker setup goo.

The Go build system, for example, just sets some GCE instance metadata:

```
        if hconf.IsContainer() {
                addMeta("gce-container-declaration", fmt.Sprintf(`spec:
  containers:
    - name: buildlet
      image: 'gcr.io/%s/%s'
      volumeMounts:
        - name: tmpfs-0
          mountPath: /workdir
      securityContext:
        privileged: true
      stdin: false
      tty: false
  restartPolicy: Always
  volumes:
    - name: tmpfs-0
      emptyDir:
        medium: Memory
`, opts.ProjectID, hconf.ContainerImage))
        }
```

Nothing about docker or docker networks or docker pulls.

And doesn't require that we vendor in new JSON and YAML parsers too. :)

I'm sure Ignition might do more, but I want less.

**Comment by bradfitz** [comment]
Your CL's shell includes:

```
        # We could have an id already but not have it on disk if e.g. the last docker
        # create blew up part-way through.
        # In that case, the only thing we can do is fall back to the
        # eventually-consistent docker daemon and hope it doesn't lie to us.
```

That is all scary and not something that should be our job.

**Comment by euank** [comment]
@bradfitz It's a fair criticism that using ignition correctly is a bit more clunky than using cloud-init.

> COS manages containers. It seems that CoreOS still manages systemd units and it's our job to do all the Docker setup goo.

I wasn't aware that COS had [konlet](https://github.com/GoogleCloudPlatform/konlet) for a simpler way of starting containers, that's cool!

Unfortunately, I think it also can't do what perkeepd wants right now. Namely, perkeepd's deployment setup wants mysql + perkeep, and for them to be linked or otherwise networked together. Konlet currently is documented as having a limitation of one container per VM. That works for the go build process, but if we want to port perkeepd's current deployment setup over, we'll end up with the exact same pile of systemd units and shell as on CoreOS.

**Comment by euank** [comment]
I guess I'm convinced enough that it's not pointless to switch from CoreOS to COS in the short term.
I think you mentioned you'd started that, @mpl; I'll hold off on my changes and then re-start on top of your work if you don't already incorporate them or make them redundant).

> @euank, we could just have one container with Perkeep + MySQL, no?

Sure, it's technically possible to do that, but  I think you  get funny looks from  other people for using docker like that :)

It would probably be easier to just use the sqlite backend instead if we're only using one host anyways.

If you want  to migrate to a 1-container konlet setup with a sqlite backend, that sounds reasonable. Is that what you're thinking, or do you want to move it to COS in whatever configuration to ease transition if/when it happens in the future?

**Comment by mpl** [comment]
> It would probably be easier to just use the sqlite backend instead if we're only using one host anyways.

yeah, I was thinking about that too (except goleveldb since that's our default).

@bradfitz any reason to stay with MySQL? I think you told me that even you gave up on it, and were running with goleveldb only now.
